spring-boot-starters/langchain4j-open-ai-spring-boot-starter/src/main/java/dev/langchain4j/openai/spring/AutoConfig.java

```diff
@@ -0,0 +1,136 @@
+package dev.langchain4j.openai.spring;
+
+import dev.langchain4j.model.openai.OpenAiChatModel;
+import dev.langchain4j.model.openai.OpenAiEmbeddingModel;
+import dev.langchain4j.model.openai.OpenAiLanguageModel;
+import dev.langchain4j.model.openai.OpenAiModerationModel;
+import org.springframework.boot.autoconfigure.AutoConfiguration;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Lazy;
+
+import static dev.langchain4j.internal.Exceptions.illegalArgument;
+import static dev.langchain4j.internal.Utils.isNullOrBlank;
+import static dev.langchain4j.openai.spring.Properties.PREFIX;
+
+@AutoConfiguration
+@EnableConfigurationProperties(Properties.class)
+public class AutoConfig {
+
+    private static final String MISSING_CONFIG_ERROR = "You are requesting an instance of an %s, " +
+            "but no '%s.%s' configuration properties are provided.";
+    private static final String MISSING_API_KEY_ERROR = "Please provide an OpenAI API key through " +
+            "the '%s.%s.api-key' property.";
+
+    @Bean
+    @Lazy
+    @ConditionalOnMissingBean
```

Comment:
A `@ConditionalOnProperty` would be easier to configure when you'll have several beans (I'm guessing you'll have them all in the same package)

Reply:
Sorry the ConditionnalBean is what they are using in Spring AI (see https://github.com/spring-projects/spring-ai/blob/1a8fcac206070990f312986f34649bda9c6ac003/spring-ai-spring-boot-autoconfigure/src/main/java/org/springframework/ai/autoconfigure/azure/openai/AzureOpenAiAutoConfiguration.java#L43C3-L43C3 ), the trick is that they split their code in one starter per technology, and that's more work but that's probably the best idea

Reply:
I am splitting into one autoconfigure/starter per provider. I see that they mix all the autoconfigurations in one module, then have starters per each providers that references autoconfigure module. Not sure what's the point?

Reply:
We're looking at the same code with @SandraAhlgrimm :-)
Not sure why they did it, then yes it should be separated in one starter per provider.

Reply:
No I would keep `@ConditionalOnMissingBean` now, so that people can create their own if they want to. Then if you split starters by providers, it's as good as a missing property.

Reply:
`@ConditionalOnProperty(PREFIX + ".chat-model.api-key")` simplifies a bit, I don't need to check if user provided properties or not. User still can provide his own `@Bean`

Reply:
Not sure I fully understand your point. Let's say user added `langchain4j-open-ai-spring-boot-starter` to their dependencies. If he/she does not provide api key, bean is not created automatically. If they provide API key, it will be created automatically. If they wish to configure their own, they remove properties and provide their own `@Bean`. Seems good, no?

Reply:
Yes that works too. That's what's happening also quite frequently with some Spring Boot starters when you need to add a key like `.enabled=true` (for example `management.endpoint.info.enabled=true`).
But with `@ConditionalOnMissingBean` it works by magic (no key), as adding the starter is enough to tell you want that feature. And then if you want to create your own bean (because there's a feature you need that's not available in the starter), you can then re-use the keys, including the API key one.

Reply:
Every starter `@Bean` that we will provide will have at least one mandatory property (e.g. API key for managed services and base URL for local deployments), so user won't be able to just add a starter dependency and get a bean automagically.

I could use `@ConditionalOnMissingBean`, but then it sohuld be either `@Lazy` or `@ConditionalOnProperty`. `langchain4j-open-ai-spring-boot-starter` provides 5 beans, if they will be just `@ConditionalOnMissingBean`, Spring Boot will attempt to create all of them and will fail because they have mandatory properties (API key).